CSV output for optimization iterations

recuero opened the pull request:
Closes #25327

recuero committed:
CSV output is outputted at each optimization iteration.

It remains to provide a fitting design to this capability.

recuero committed:
Add interval option for CSV and test (#25327)

Check iteration output for one combined module test.

recuero committed:
Move csv tests to output folder

moosebuild commented:
Job [Coverage](https://civet.inl.gov/job/1732510/) on 4e89aa5 wanted to post the following:


## Framework coverage

<table>
  <thead>
    <tr>
      <th></th>
      <th>516ad1</th>
      <th colspan=4>#25328 4e89aa</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>85.64%</td>
      <td>85.64%</td>
      <td>-0.00%</td>
      <td>58.82%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>90353</td>
      <td>90361</td>
      <td>+8</td>
      <td>10</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>15147</td>
      <td>15154</td>
      <td>+7</td>
      <td>7</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25328/coverage_diff/framework)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25328/coverage/framework)


## Modules coverage

### Optimization

<table>
  <thead>
    <tr>
      <th></th>
      <th>516ad1</th>
      <th colspan=4>#25328 4e89aa</th>
    </tr>
    <tr>
      <th></th>
      <th>Total</th>
      <th>Total</th>
      <th>+/-</th>
      <th>New</th>
    </tr>
  </thead>
  <tbody>
    <tr>
      <td><b>Rate</b></td>
      <td>89.77%</td>
      <td>89.79%</td>
      <td>+0.02%</td>
      <td>100.00%</td>
    </tr>
    <tr>
      <td><b>Hits</b></td>
      <td>1351</td>
      <td>1354</td>
      <td>+3</td>
      <td>3</td>
    </tr>
    <tr>
      <td><b>Misses</b></td>
      <td>154</td>
      <td>154</td>
      <td>-</td>
      <td>0</td>
    </tr>
  </tbody>
</table>

[Diff coverage report](https://mooseframework.inl.gov/docs/PRs/25328/coverage_diff/optimization)

[Full coverage report](https://mooseframework.inl.gov/docs/PRs/25328/coverage/optimization)

## Full coverage reports

<details>
<summary>Reports</summary>

- [`framework`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/framework)
- [`chemical_reactions`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/chemical_reactions)
- [`combined`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/combined)
- [`contact`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/contact)
- [`electromagnetics`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/electromagnetics)
- [`external_petsc_solver`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/external_petsc_solver)
- [`fluid_properties`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/fluid_properties)
- [`fsi`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/fsi)
- [`functional_expansion_tools`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/functional_expansion_tools)
- [`geochemistry`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/geochemistry)
- [`heat_conduction`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/heat_conduction)
- [`level_set`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/level_set)
- [`misc`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/misc)
- [`navier_stokes`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/navier_stokes)
- [`optimization`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/optimization)
- [`peridynamics`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/peridynamics)
- [`phase_field`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/phase_field)
- [`porous_flow`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/porous_flow)
- [`ray_tracing`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/ray_tracing)
- [`rdg`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/rdg)
- [`reactor`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/reactor)
- [`richards`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/richards)
- [`scalar_transport`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/scalar_transport)
- [`solid_properties`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/solid_properties)
- [`stochastic_tools`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/stochastic_tools)
- [`tensor_mechanics`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/tensor_mechanics)
- [`thermal_hydraulics`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/thermal_hydraulics)
- [`xfem`](https://mooseframework.inl.gov/docs/PRs/25328/coverage/xfem)

</details>

## Warnings

- `framework` new line coverage rate 58.82% is less than the suggested 90.0%



This comment will be updated on new commits.

recuero committed:
Small fixes

GiudGiud commented on framework/include/executioners/Steady.h:
```diff
@@ -56,12 +56,31 @@ class Steady : public Executioner
+   *
+   * @param is_optimize_solve Whether the system is performing optimization.
+   */
+  virtual void setOptimizationFlag(const bool is_optimize_solve)
```
We need to be able to not have that

I know you are trying to follow previous guidance but you found a way around this (no optim-specific code in framework, instead stuff usable for other purposes as well) for exodus and we need to be using the same mechanic

recuero committed:
Rename API for generality.

Same functionality could be used for outputting per-iteration
solution variables of other solution procedures, not just
optimization solves.

Addresses Guillaume's review.

lindsayad commented on framework/src/outputs/CSV.C:
```diff
@@ -173,6 +174,15 @@ CSV::getVectorPostprocessorFileName(const std::string & vpp_name,
                 << std::right << _linear_iter;
     }
+
+    // These lines enable CSV per-optimization iteration output (Optimization module)
+    if (auto steady = dynamic_cast<Steady *>(_app.getExecutioner()))
+      if (steady->getIterationOutputFlag() && _current_execute_flag == EXEC_TIMESTEP_END)
+      {
+        std::string append = "iteration_" + Moose::stringify(steady->getIterationNumberOutput());
+        file_name << '_' << std::setw(_padding) << std::setprecision(0) << std::setfill('0')
+                  << std::right << append;
+      }
```
I think we're crossing over the red line here. It seems like at least some of these APIs can go on the base executioner?

recuero replied:
We can do that to avoid casting. But I don't know that we have a use case now.

lindsayad replied:
It seems like we are realizing @dschwen's fears in https://github.com/idaholab/moose/pull/25011#issuecomment-1650815614. We are somehow in the pretty undesirable middle ground between having kept logic internal to the module vs. having a sufficiently abstract design that it makes sense in the base class

lindsayad reviewed (changes requested):
It would be nice to not see any dynamic casts

recuero committed:
Move per-iteration output API to base executioner class.

Addresses Alex's review.

lindsayad commented on framework/src/outputs/CSV.C:
```diff
@@ -173,6 +174,17 @@ CSV::getVectorPostprocessorFileName(const std::string & vpp_name,
                 << std::right << _linear_iter;
     }
+
+    // These lines enable CSV per-optimization iteration output (Optimization module)
```
```suggestion
    // These lines enable things like per-optimization iteration output (Optimization module)
```

lindsayad commented on framework/src/outputs/CSV.C:
```diff
@@ -173,6 +174,17 @@ CSV::getVectorPostprocessorFileName(const std::string & vpp_name,
     }
+
+    // These lines enable CSV per-optimization iteration output (Optimization module)
+
```
```suggestion
```

lindsayad commented on framework/src/outputs/Output.C:
```diff
@@ -25,6 +25,7 @@
 #include "Function.h"
 #include "PiecewiseLinear.h"
 #include "Times.h"
+#include "Steady.h"
```
```suggestion
```

lindsayad commented on framework/include/outputs/Output.h:
```diff
@@ -165,6 +165,11 @@ class Output : public MooseObject,
+  /**
+   * Get the time step or optimization (module) iteration number for interval output.
+   */
+  virtual int getTimeStep();
```
```suggestion
  virtual int getStep();
```

lindsayad commented on framework/src/outputs/CSV.C:
```diff
@@ -11,6 +11,7 @@
 #include "CSV.h"
 #include "FEProblem.h"
 #include "MooseApp.h"
+#include "Steady.h"
```
```suggestion
```

lindsayad commented on framework/include/executioners/Executioner.h:
```diff
@@ -153,6 +153,42 @@ class Executioner : public MooseObject,
+   * nested solve. This output iteration number may be set by the parent app for a sub-app. This
+   * behavior is decided by the Executioner/Executor/SolveObject in charge of the solve.
+   */
+  virtual unsigned int getIterationNumberOutput() const { return _output_iteration_number; }
```
I wonder if this should take an execute flag key. Then we could think about nesting iteration number output, e.g. what iteration number am I on in my `EXEC_NONLINEAR` loop? And at the same time, what iteration number am I on in my `EXEC_LINEAR` loop? Or alternatively there is no key, but we return some kind of ordered map which is a sequence of execute flags and their iteration numbers

recuero replied:
>  Then we could think about nesting iteration number output, e.g. what iteration number am I on in my `EXEC_NONLINEAR` loop?

That's a type of behavior that opens up with optimization (or other outer iteration) type of problems. I think that that type of output is less relevant for optimization problems than the result of the forward-adjoint solve (the iteration number doesn't change as you solve physics problems, so you'd just end up with normal system-wide convergence output which can be obtained from a non-optimization solve). It also seems that nothing gets updated in the subapps (e.g. forward) on nonlinear or linear, even if the flags are set in the postprocessors/output blocks. I.e. I am getting no additional output as I enable linear/nonlinear "outer iteration" output on `CSV.C`.  

The flags we could set here from the optimization module (`OptimizeSolve`) are optimization-specific flags (`EXEC_FORWARD`, `EXEC_ADJOINT`, `EXEC_ADJOINT_TIMESTEP_BEGIN`,...).

lindsayad replied:
> linear/nonlinear "outer iteration"

Wouldn't these be "inner" iterations?

recuero replied:
Yes.

lindsayad replied:
My concern with the current API, which is of course a product of us pushing you to generalize, is that its unclear what `getIterationNumberOutput()` means on its own. Does it mean the time step number? The nonlinear iteration number? The linear iteration number? The adaptivity step in a grid refinement study of a steady solve? This is why I'd propose APIs like this

```c++
virtual const std::vector<std::pair<ExecFlagType, IterationInfo>> & getIterationInfo() const;
virtual const IterationInfo & getIterationNumber(const ExecFlagType &) const;
```

where `IterationInfo` is a structure whose data members we could brainstorm. Certainly a `step` or `iteration_number` would be one necessary data member.

What are your thoughts? Maybe we should have a Teams call with @GiudGiud and anyone else who would be interested. Perhaps @dschwen

recuero replied:
Using data structures to better describe the output is a good idea. Separately, but related, I think we could fall into similar "contradictions". For example, if `step` or `iteration_number` are data members: What kind of iteration is captured with the `iteration_number` member? Is it an optimization one (we don't want to say that explicitly in the framework)? It seems as if we want _special_ output for an application that, from the framework, we know nothing of. This type of special output would probably beg for child optimization module objects.

We can talk more offline. I can also use your `IterationInfo` approach above.

lindsayad replied:
The IterationInfo is paired with an execution flag in the APIs I proposed. So there would hopefully be no ambiguity. 

I could definitely see a future though where IterationInfo (or whatever it ended up being) is a class that can be derived from. I don’t know if the future is here yet

recuero replied:
An execution flag in the framework does not necessarily define the iteration type we are after. We are using here TIMESTEP_END because it works, but it can still be equivocal. I'm not opposed to doing that, but I believe the design issue wouldn't be resolved with that data structure.

lindsayad replied:
Applications can register their own execution flags as the optimization module does

lindsayad commented on framework/src/outputs/CSV.C:
```diff
@@ -173,6 +174,17 @@ CSV::getVectorPostprocessorFileName(const std::string & vpp_name,
+
+    // These lines enable CSV per-optimization iteration output (Optimization module)
+
+    if (_app.getExecutioner()->getIterationOutputFlag() &&
+        _current_execute_flag == EXEC_TIMESTEP_END)
```
Does the optimization module have its own execute on flags?

recuero replied:
Yes, it has its own execute on flags: `EXEC_FORWARD`, `EXEC_ADJOINT`, `EXEC_ADJOINT_TIMESTEP_BEGIN`,...

recuero committed:
More cleanup (Alex's review)

recuero commented:
@lindsayad will investigate having more general output capabilities.
